### ISSUE TITLE
Reproducing results in paper

### ISSUE THREAD

**Issue by mmmiiinnnggg** [usamec/double_binary#3]
Hi guys, I found it's a very interesting method to compress LLM in low-bit with scalar quantization, and I performed experiments to try to replicate your results. 

### TL; DR
It didn't worked out as expected. This is a gap between my reproduction and paper claimed results (particularly the performance after compression)

### Hardware/libraries versions
torch==2.3.1
transformers==4.43.3

I used nvidia-v100 gpus. Since Volta does not support originally the BF16 computation, I use `torch.float32`. The `BF16FusedAdamW` was replaced by regular AdamW optimizer from torch, and `linear_cross_entropy` was replaced by regular `nn.CrossEntropyLoss()` since I didn't encounter the OOM problem.

### for llama2-7b model
The tested target bit-width is 1.5 BPW. Since the code in the repository is only for llama3-8b, I didn't changed them and just inherited from it. Specifically: 
- **for compression**: `seq_len = 4096`, `num_calib_data=256`, `lr=3e-5`, `weight_decay=1e-4`, `batch_size=8`, `training_epoch=8`. The regularization term for ADMM algorithm is set to `3e-2`, with iteration number = 3. The outer alternating optimization step = 260.
- **for finetuning**: train scaling vectors using KD with 10000 iteration. PV-tuning is disabled.

|  | PPL  after compression | PPL after finetuning | Winograde|PiQA|Arc Easy|Arc Challenge| 
|----------|----------|----------|----------|----------|----------|----------|
| Paper claimed  | -    | 7.16   |64.4|73.3|66.3|35.1|
| Reproduction    | 18.96     | 7.54   |63.2|72.6|63.6|32.7|

### for llama3-8b model
The tested target bit-width is 2.0 BPW. All hyperparameters are taken from your shared notebook,
Specifically: 
- **for compression**: `seq_len = 8192`, `num_calib_data=256`, `lr=3e-5`, `weight_decay=1e-4`, `batch_size=8`, `training_epoch=8`. The regularization term for ADMM algorithm is set to `3e-2`, with iteration number = 3. The outer alternating optimization step = 260.
- **for finetuning**: train scaling vectors using KD with 10000 iteration. PV-tuning is disabled.

|  | PPL  after compression | PPL after finetuning | Winograde|PiQA|Arc Easy|Arc Challenge| 
|----------|----------|----------| ----------|----------|----------|----------|
| Paper claimed  | 9.05 ?(The number of last cell output of notebook)    | 7.48   |68.0 | 76.8|74.5|41.0|
| Reproduction    | 14.17     | 8.14   |67.2|76.0|71.2|38.2|

### Conclusion
The finetuning stage seems work fine, but it seems that the performance gap after compression is particularly significant, which hurts the starting checkpoint for finetuning and thus final results. 

Could you give us some insights about what could be the missing ingredients? The deviation of results after compression seems not to be the reason, since for llama2 model I run several times with deviation of PPL around 3.

**Comment by usamec** [comment]
I agree, this is significant and ugly gap :/

I pushed Llama2-7B notebook for 1.5 BPW, it has ~11 PPL compared to yours ~18. Same happens for Llama3-8B.

So, I would look for the difference during the compression phase. Can you please, somewhere share your compression logs/notebooks?

**Comment by usamec** [comment]
PV-tuned checkpoints are here:
http://compbio.fmph.uniba.sk/~usama/DBF/latest_llama2-1b-15fin.pt
http://compbio.fmph.uniba.sk/~usama/DBF/latest_llama3-1b-20fin.pt

**Comment by usamec** [comment]
One more possibility: I noticed in other experiments that, if your RedPajama contains some bad sample (something which model thinks is totally improbable), it can throw the whole output importance gradient estimation off. 

Using true fisher is more robust to this outliers and gives slightly better results:
Notebook with it here: https://github.com/usamec/double_binary/blob/truefisher/raw_stuff/compress/Compress-Llama-3-8B-tar20-truefisher.ipynb but the only changed part is cell 6 which norms estimation.

**Comment by mmmiiinnnggg** [comment]
> I agree, this is significant and ugly gap :/
> 
> I pushed Llama2-7B notebook for 1.5 BPW, it has ~11 PPL compared to yours ~18. Same happens for Llama3-8B.
> 
> So, I would look for the difference during the compression phase. Can you please, somewhere share your compression logs/notebooks?

Hi, this is my code for reproduction: https://github.com/mmmiiinnnggg/dbf_reproduce/tree/main. I refactored so that it is easier to adjust hyperparameters.

**Comment by mmmiiinnnggg** [comment]
Thanks for your shared materials! I will investigate into it. But I also have some questions about the method itself:
1. What is the amount of finetuning data? In your code it seems to load the whole tokenized redpajamas dataset (which contains ~1T tokens in total) and perform finetuning with DDP, but according to your paper, the finetuning process can be finished with 4 A100 in a day, which seems only a portion of whole dataset is used.
2.  What is the insights of choosing specifically q, v, o matrices to factorize firstly in compression, and then finetune other matrices and then factorize? Could we choose other matrices for first round of factorization in each block, or maybe factorize all matices at once?

**Comment by usamec** [comment]
I will check your code later, thanks.

As for the questions:
"What is the amount of finetuning data?" - for compression it is just 256*8192 tokens (or 256*4096 for llama2). 
For fine-tune it is just 1B redpajama sample (same as used in PV-tuning paper, see https://github.com/Vahe1994/AQLM?tab=readme-ov-file#preparing-fine-tuning-dataset which uses togethercomputer/RedPajama-Data-1T-Sample) and in reality it is even less (there is 1 day limit in our cluster and we get only over like 0.5B tokens during that time).

"What is the insights of choosing specifically q, v, o matrices to factorize firstly in compression, and then finetune other matrices and then factorize? Could we choose other matrices for first round of factorization in each block, or maybe factorize all matices at once?"

This was kind of arbitrary. I wanted to be able to compensate for some errors using nearest matrices. The main gain was mainly by fine-tuning towards the original block output (in the original non-compressed network), not toward block output with compressed input from previous blocks.

**Comment by mmmiiinnnggg** [comment]
<img width="408" height="507" alt="Image" src="https://github.com/user-attachments/assets/76ae9d2f-06d7-4db0-80be-cae5a24b293a" />

One thing I noticed very strange is that the error magnitude after each epoch in compression in my code is high. But in your case the error magnitude always maintain at the level of 1e-2.

But the when training matrices in first block, seems fine (the magnitude ~ 1e-4)

**Comment by usamec** [comment]
Where can I find from apps.llm.finetuning.utils.data import HuggingFaceDataloader ?

**Comment by mmmiiinnnggg** [comment]
> Where can I find from apps.llm.finetuning.utils.data import HuggingFaceDataloader ?

Oh, it is in my another codebase, which is used everywhere. I use this because I have already have tokenized redpajams/wikitext which is loaded through this class, so I want to reuse it. I used this class to load the wiki2 data and tested the checkpoint that you shared (PV-tuned llama2-7b-1.5BPW) and get 7.01 PPL as claimed in the paper. So I think maybe the problem is not there. 

If you really need this class I can share it later because right now I don't have my laptop with me ;(

**Comment by usamec** [comment]
I replaced it, no problem :)

But I noticed one thing. You load model as bfloat16 (https://github.com/mmmiiinnnggg/dbf_reproduce/blob/main/dbf_utils.py#L24), and never change it to float32 before entering optimization loop here https://github.com/mmmiiinnnggg/dbf_reproduce/blob/main/dbf_utils.py#L535

Which means you use AdamW with bfloat16. That leads to precision issues (i.e. in bfloat16 1+0.001 = 1), that's why I had BF16FusedAdawW there (which adds additional mantissa for weights as optimizer state).
But you mentioned, that you use float32, so is this just wrong code pushed?

Also, I would suggest keeping err_before block there, because occasionally in the first couple of layers error blows up and this helps catch it.

**Comment by mmmiiinnnggg** [comment]
> I replaced it, no problem :)
> 
> But I noticed one thing. You load model as bfloat16 (https://github.com/mmmiiinnnggg/dbf_reproduce/blob/main/dbf_utils.py#L24), and never change it to float32 before entering optimization loop here https://github.com/mmmiiinnnggg/dbf_reproduce/blob/main/dbf_utils.py#L535
> 
> Which means you use AdamW with bfloat16. That leads to precision issues (i.e. in bfloat16 1+0.001 = 1), that's why I had BF16FusedAdawW there (which adds additional mantissa for weights as optimizer state). But you mentioned, that you use float32, so is this just wrong code pushed?
> 
> Also, I would suggest keeping err_before block there, because occasionally in the first couple of layers error blows up and this helps catch it.

Oh, I see, maybe that's the root cause... I checked my code again, I indeed used the bfloat16 all the time, because on V100 32GB there was some problem of fitting model and perform computation in FP32.  I rerunned compression at A800 again, hoping natively supported bf16 could help, but it still didn't reproduce. So the claimed usage of "fp32" is actually wrong ;)

Anyway, I will try to run experiment in real FP32, or install `bf16_fused_adam`. I will come back with verified results.

**Comment by usamec** [comment]
Great. Also, please push the logs from it somewhere.

**Comment by usamec** [comment]
And http://compbio.fmph.uniba.sk/~usama/DBF/rp2.zip should contain calibration/finetuning data for Llama2.

**Comment by mmmiiinnnggg** [comment]
Hi, I reruned the llama2-7b experiments with float32 computation. But PPL still didn't reach to ~11. Here are the details:

1. load model as float32: PPL = 22.19
2. load model as bfloat16, then cast to float32 before entering optimization loop: PPL = 17.04

I pushed new `dbf_utils.py` in my reproduction repo, where the change is layer.float() - https://github.com/mmmiiinnnggg/dbf_reproduce/blob/0973fc0798635870b012eef38a9e2a3a072c665e/dbf_utils.py#L499. I printed out the dtype of trainable parameters passed to optimizers https://github.com/mmmiiinnnggg/dbf_reproduce/blob/0973fc0798635870b012eef38a9e2a3a072c665e/dbf_utils.py#L524.
I pushed log information of layer-wise compression when running experiments using this script. 
https://github.com/mmmiiinnnggg/dbf_reproduce/blob/main/llama2-7b-tar15-0_15_blocks.txt
https://github.com/mmmiiinnnggg/dbf_reproduce/blob/main/llama2-7b-tar15-16_31_blocks.txt

Right now I am running experiments with your original script and data that you shared to verify the training data difference. (with only modification of load model in float32, and use AdamW optimizer from torch)
- "Also, I would suggest keeping err_before block there, because occasionally in the first couple of layers error blows up and this helps catch it." what should be done when `err before` blows up? I have occurred with your code and your data. This happened when training weights in layer.1
<img width="1480" height="193" alt="Image" src="https://github.com/user-attachments/assets/36a259a7-37d1-4ee3-8e6a-664defbb2e3b" />

UPD: the experiment with your script and data gets PPL=14.97/16.54 (run2)/13.14(using cosine scheduler) for llama2-7b-1.5BPW. But it is noticeable that some `err_before`  blowed up and have big difference when comparing output logs. Very much of needing suggestions to avoid err_before blowing up..

**Comment by mmmiiinnnggg** [comment]
I noticed one thing: when calculating loss, you multiply by `layer.mlp.down_proj.o_norm`. Is this could be the possible reason because of bad estimated gradient norm? Also, you multiply a big number (10^6) at backward hook https://github.com/mmmiiinnnggg/dbf_reproduce/blob/0973fc0798635870b012eef38a9e2a3a072c665e/dbf_utils.py#L370, it there some insights about this multipier?

**Comment by usamec** [comment]
The o_norm multiplication is just an importance weighting of output neurons (gives slightly better results, but removing it, should not change a lot).

The big number is just to avoid 10^-9 losses and similar. 

I am reruning full code right now, so I should have some insights tmrw.

As for avoiding the blowup. Try lowering the learning rate a bit (maybe only for the first two blocks).

**Comment by usamec** [comment]
Run your script with couple of changes, got: "Got PPL 10.50129935240124". This was on A100, compression runs in my repo are on 4090.
See commit here https://github.com/usamec/dbf_reproduce/commit/148ada090d626bb8cd89714f14833513ec1721db

Also, did you try your Wikitext2 dataloader with uncompressed Llama and get 5.12 PPL on Llama2?

**Comment by mmmiiinnnggg** [comment]
> Run your script with couple of changes, got: "Got PPL 10.50129935240124". This was on A100, compression runs in my repo are on 4090. See commit here [usamec/dbf_reproduce@148ada0](https://github.com/usamec/dbf_reproduce/commit/148ada090d626bb8cd89714f14833513ec1721db)
> 
> Also, did you try your Wikitext2 dataloader with uncompressed Llama and get 5.12 PPL on Llama2?

Thanks for your reply!

Yes, I tried with my own dataloader to test llama2 with seqlen=4096 and got 5.11638

Could you please verify that if load model in float32 from the beginning and use ordinary `torch.optim.AdamW` instead of `BF16FusedAdamW`, what will be the results? That seems the single difference that I can tell from your commit. I will also check it from my side.

**Comment by usamec** [comment]
Got OOM killed in the middle of fp32 run :/,  so it will take some time.

**Comment by mmmiiinnnggg** [comment]
FYI: runned two experiments with ordinary `torch.optim.AdamW` and `float32` (disabled attn_implementation of flash-attn-2) with your commit code (train data from your tokenized redpajama). Experiment was done at V100.
Got PPL 13.307512091971345
Got PPL 15.673891793584179

Observed again `err_before` blowed up at several blocks (train err > 2 or more)

Wait for your float32 results, but seems the only option left would be align with your torch and transformers version, install BF16FusedAdamW and run experiments at A100/4090 with bf16 supported? (which seems odd since fp32 covers the precision of bf16)

**Comment by usamec** [comment]
My fp32 got PPL 10.132203069569947. https://github.com/usamec/dbf_reproduce/blob/main/run32-2.txt

I also pushed requirements for the whole env I used (a lot of them are irrelevant sorry about that) https://github.com/usamec/dbf_reproduce/blob/main/requirements.txt

**Comment by mmmiiinnnggg** [comment]
Ok, then it should be some libraries issue since my code is identical to yours.

Thanks for your pragmatic and quick responce! I do believe it's a valuable method and that's why I am trying to figure it out.

**Comment by mmmiiinnnggg** [comment]
Hi! After aligning with your torch and transformers version (2.7 and 4.53, my previous env was 2.3 and 4.43), I reproduced results:
PPL=10.202370067820578
PPL=10.231628249544992
That's quite unexpected but reasonable, since we excluded all other possible reasons. It's quite interesting why changing libraries version could affect results a lot, although the versions differences are not significant in my opinion. I would investigate into it, but thanks for your responses all the time.) 

P.S. Another thing I have noticed is that my dataloader take 256 samples each time differently, while you take data samples and fixed for collecting norms and optimizing blocks.

**Comment by mmmiiinnnggg** [comment]
P.P.S. I conducted more ablation experiments of comparing env and dataloader, and the conclusion is that the problem is related to my DATALOADER, which samples different calibration method for different block. This means we use different data for different block. In your implementation (and in general PTQ) the calibration dataset is sampled before optimization and fixed. So the really root cause is my train dataloader. I finally debugged. Thanks!

**Comment by usamec** [comment]
Glad to see, that it is fixed. 

Closing this then :)
